server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocatorTests.java

```diff
@@ -401,15 +405,56 @@ public void onFailure(Exception e) {
         }
     }
 
-    private static DiscoveryNode createDiscoveryNode(String nodeId) {
-        return new DiscoveryNode(
-            nodeId,
-            nodeId,
-            buildNewFakeTransportAddress(),
-            Map.of(),
-            Set.of(DiscoveryNodeRole.MASTER_ROLE, DiscoveryNodeRole.DATA_ROLE),
-            Version.CURRENT
+    public void testResetDesiredBalance() throws InterruptedException {
+
+        var node1 = newNode(LOCAL_NODE_ID);
+        var node2 = newNode(OTHER_NODE_ID);
+        var clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .nodes(DiscoveryNodes.builder().add(node1).add(node2).localNodeId(node1.getId()).masterNodeId(node1.getId()))
+            .build();
+
+        var threadPool = new TestThreadPool(getTestName());
+        var clusterService = ClusterServiceUtils.createClusterService(clusterState, threadPool);
+        var desiredBalanceComputer = mock(DesiredBalanceComputer.class);
```

Comment:
I'd rather we did this without mocks too. We can just subclass the real implementation I think.

Reply:
Why? Subclass would lead to extra complexity of creating all the constructor dependencies

Reply:
I've spent enough time hunting down sneaky NPEs from refactorings involving mocks to last a lifetime already :)

There's only 3 dependencies: a `ClusterSettings` which is trivial to create, a `ThreadPool` that we already have, and a `ShardsAllocator` that can be null.